Chainparams: Generic chainparam selection with -chain=<strNetworkID>

jtimon opened the pull request:
Continues #4804.
Deprecate -testnet and -regtest in favor of -network=testnet and -network=regtest.
When the old args are used a warning will be logged.
The repetition of the invalid combination error string is also avoided.
Additionally, the chainparam networkID is removed since it's not used anymore (as it shouldn't).

jtimon committed:
Chainparams: The hash of the genesis block it's the genesis checkpoint and chain id

jtimon commented:
Closing until #5598 gets merged.

jtimon commented:
Reopened and rebased after #5598 has been merged. Now it's easier to review.

petertodd commented:
Rather that having the argument be "-network" why don't we rename it to "-chain"?

Network is a term that's already overloaded in multiple contexts. After all, we called it *chain* params rather than network params.

On 7 January 2015 18:48:08 GMT-05:00, "Jorge Timón" <notifications@github.com> wrote:
>Reopened and rebased after #5598 has been merged. Now it's easier to
>review.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/bitcoin/bitcoin/pull/5229#issuecomment-69113620

jtimon commented:
Yes, I also thought about "mode" and "chain". Technically main and unittest are the same chain, since they share the same genesis block and all, so I decided for "network". But that's the only exception so probably "chain" still makes sense.
Let's wait for other people to decide the color of the bike shed and I'll make the change to whatever is more popular.

jtimon commented:
Needed rebase so I squashed the nits.

laanwj commented:
I'm still not convinced that this is necessary. The reason for this change seems very academic; `-testnet` `-regtest` versus `-network=testnet` `-network=regtest`, meh.

petertodd commented:
@laanwj FWIW I always use --testnet and --regtest in my command-line tools.

laanwj commented on src/chainparamsbase.cpp:
```diff
@@ -94,22 +93,26 @@ CBaseChainParams::Network NetworkIdFromCommandLine()
+
+    std::string chain = GetArg("-chain", "main");
+    if (chain == "main")
+        return CBaseChainParams::MAIN;
```
What if these would be defined as string constants, e.g.:
```
    const std::string CBaseChainParams::MAIN = "main";
    const std::string CBaseChainParams::TESTNET = "testnet";
    const std::string CBaseChainParams::REGTEST = "regtest";
```
... I guess we could get rid of the `CBaseChainParams::Network` enumeration type completely, making `SelectParams` take a string. This kind of switching code would be no longer necessary apart from the data structure selection in SelectParams itself.

jtimon replied:
Well, we can get rid of the enumeration even if the strings aren't declared as constants. I won't opposed to it but working at big java projects were those type of constants are abundant I have always wondered if the gain in readability that CBaseChainParams::TESTNET provides compared to "testnet".
Yes, yes, developers could type "testtnet" and have some fun getting an "unkown chain 'testtnet'" exception...

Personally I like the Params(type) or Params(std::string) factory more than Params() in combination with SelectParams(). I mean, this is closer to what I would do: https://github.com/jtimon/bitcoin/commit/d135001a0a8ae9536553d49b1ac8a3c2e0643e41#diff-d22bc3e058f8982972e2eb381a1df668R34
In fact my preference when it comes to creating polymorphic objects would be for the factory function to return a pointer to a newly created object that the caller has to delete, a good-old factory without singleton [like Params()] or per-type singleton [like Params(std::string)], but that's more than we need. 
I was scared that moving from enum to string was going to be something people wouldn't like, but if that's not the case I can definitely improve this.
The remaining question is, do we really need the string constants when you will get an exception right in the face if you don't use the correct string?
In my opinion, "the same string you use in command line and it's documented" would be ideal.

laanwj commented:
This also needs changes to the qt autostart functionality, #5793, which relies on passing through -regtest/-testnet at the moment. But at least that is a simplification, you could just pass `-chain=%s` with %s the current chain name.

jtimon commented:
@laanwj could you give a line number link on where is that passed?

jtimon commented:
@petertodd as discussed those options will remain usable and not deprecated.
Given that unittestnet has disappeared, it's really mainly an extensibility and altchain friendly patch so I understand if it remains in limbo until it is more directly useful for bitcoin.
It won't be hard to maintain unmerged anyway. Whenever somebody thinks about adding a third --harforktestnet or whatever this PR will be here to remind us that N is better than 2 + 1 (+ the implicit default "main" or "bitcoin").
We can label it low priority or I can put an uppercased word at the beginning of the tittle to clarify that people shouldn't get very distracted by this unless there's nothing else to merge or the somebody reanimates it [by proposing an additional chainparam option in the command line].
Paint it black and let it be, or something.
I can add a compilation error for a red cross to be shown.
But keeping it open I remember to rebase it and "police" this part of the code in some sense.
I can also just set an alarm to rebase it with the PR closed.
Not a strong opinion besides "if a chainparam option has to be added, it must be generic".
Sorry for the wall of text.

jtimon commented:
Updated with @laanwj 's suggestion for replacing the enum with strings and adding a new GetParamsHelpMessages() function. Note that it could be used in src/bitcoin-cli.cpp too but I wasn't sure if we wanted that depending on ui_interface.

jtimon commented:
Sorry, I forgot to deal with the qt autostart functionality as pointed out by@laanwj.
It is done now.

I also stopped using the bip70 nomenclature for the string constants since that changes "testnet" for just "test", which wouldn't have worked very well with the qt part.
Is there any possibility to correct bip70 in that sense?

laanwj commented on src/qt/guiutil.cpp:
```diff
@@ -706,16 +704,15 @@ bool SetStartOnSystemStartup(bool fAutoStart)
-        optionFile << "Exec=" << pszExePath << strprintf(" -min -testnet=%d -regtest=%d\n", GetBoolArg("-testnet", false), GetBoolArg("-regtest", false));
+        else
+            optionFile << strprintf("Name=Bitcoin (%s)\n", chain);
+        optionFile << "Exec=" << pszExePath << strprintf(" -min -chain=%s -regtest=%d\n", chain);
```
Spurious `-regtest=%d`

jtimon commented:
Oopss, thanks @laanwj , fixed.
Also adding a commit in which src/bitcoin-cli.cpp also uses GetParamsHelpMessages() [and also ui_interface].
Thoughts on that commit? @theuni is there an specific reason for bitcoin-cli not to use ui_interface that I'm missing?

jtimon commented:
I mean, the last commit would be kind of undoing https://github.com/bitcoin/bitcoin/commit/73ac7abd08a70adf22e24d0f5f7631e7f8b7c5bb

That's why I'm assuming people may not lie the idea.

jtimon commented:
Never mind about the SQUASHME commits, I rebased on top of #6022, which is what I think makes most sense.

jtimon commented:
Closing until #6022 is merged.

jtimon commented:
Rebased after https://github.com/jtimon/bitcoin/commit/b74dcb3b4aa29958d22135f2d6c015578069e83c ( #6022  )

jtimon commented:
Needed rebase

jtimon committed:
Chainparams: Use a regular factory for creating chainparams

jtimon commented:
Added 3 commits to hopefully make it more interesting.
The added tests are kind of redundant, but won't hurt.
I'm specially interested on @theuni 's opinion on https://github.com/jtimon/bitcoin/commit/65e2a722d64454378bb481afab154e4c9e1178d9 (since I know he wants to eventually separate the state-related functions out, but ParamsFactory() doesn't rely on any state).

jtimon commented:
Removed https://github.com/jtimon/bitcoin/commit/ac1c555f275109582974ab0245a0d81854cf1ddd since is now a proposal for the trivial branch (https://github.com/theuni/bitcoin/pull/27).

To better read commit "Chainparams: The hash of the genesis block it's the genesis checkpoint and chain id", please rgrep hashGenesisBlock

All mentions of it are whether to:

- Its own asignments, of course.
- Make sure the genesis block is in a block index we're maintaining, like in ./init.cpp:1112, of course.
- Asserts in chainparams.cpp (no need to aditionally check genesis.hashMerkleRoot), let's keep this, why not?

- Avoid unnecessary checkpoints for the genesis block, in case we want to apply what's inside it somewhere, for example: ./main.cpp:1708, ./main.cpp:2641, ./main.cpp:2732, ./main.cpp:3385, ./main.cpp:3400, ./main.cpp:3495

Why the heck are we trying to check the genesis block in the first place?
The genesis block is not something to test, it is actually the first rule of the chain, it has to be right or the chain rules are broken (softforks and hardforks will come at a later nHeight).

But what about checkpoints?

The logic of checkpoints can always safely rely on the first checkpoint being true.
The last point may not be true if three's a reorg behind it, otherwise you're ready to optimize.
We could introduce a configurable third layer of "this is my new genesis no matter what", which would be good for pruning and skipping all validations on blocks that are too old to believe they can be reorg-ed, but this would only really make sense with some form of utxo commitment (did I say that append-only-txo + append-only-confirmed is enough for this?).

Therefore, we should not:

- Disperse hardcoded values.
- Double-check hashes of hashes
- Leave checkpoints without the only true first safe checkpoint: the genesis block

jtimon commented:
Sorry, pushed too fast. Now I use containers instead of those ugly ClearSelectedParams functions, and the first commit actually compiles (now all of them do).

jtimon committed:
Chainparams: Generic selection with -chain=<chainString> in addition of -testnet and -regtest

jtimon committed:
Change?: Chainparams: DRY: Make qt/guiutil.cpp fit BIP70 chain name strings

jtimon committed:
Chainparams: Replace CBaseChainParams::Network enum with string constants (suggested by Wladimir)

jtimon committed:
Change?: bip70 CBaseChainParams::TESTNET = from "test" to "testnet3" (and user-facing testnet to "testnet3")

jtimon committed:
DRY: TODO find out why this excludes regtest

jtimon committed:
Chainparams: Translations: DRY: options strings

jtimon committed:
fixup! Chainparams: Translations: DRY: error strings

jtimon committed:
fixup! RENAME: strNetworkID = CBaseChainParams::REGTEST, strDataDir = CBaseChainParams::REGTEST

s/"regtest"/CBaseChainParams::REGTEST
s/"main"/CBaseChainParams::MAIN
s/"testnet"/CBaseChainParams::TESTNET

jtimon commented:
Closing for now, I will open a version with some of the changes but without adding the new option.

jtimon commented:
This is the promised version with some of the improvements but without adding the new option: #6229
Opened for discussion.
